sms_alternative_provider/models/ir_sms_gateway.py

```diff
@@ -37,94 +39,158 @@ def _compute_description(self):
+
             if handle_results:
-                provider._handle_results(messages_to_send, provider_result)
+                gateway._handle_results(messages_to_send, provider_result)
             result.extend(provider_result)
+
         return result
 
     @api.model
     def _send_get_providers(self, messages):
         """
         Return all providers potentially suitable for the current context
         """
-        return self.search(
-            ["|", ("company_id", "=", False), ("company_id", "=", self.env.company.id)]
+        providers = self.search(
+            [
+                ("active", "=", True),
+                "|",
+                ("company_id", "=", False),
+                ("company_id", "=", self.env.company.id),
+            ]
+        )
+        _logger.info(
+            "_send_get_providers found: %s",
+            [(p.id, p.name, p.gateway_type, p.active) for p in providers],
         )
+        return providers
 
     def _send_partition_providers(self, messages):
         """
         Return a dict with providers in self as keys and lists of messages a provider
         will handle as value
-        ie
-        {
-            ir.gateway.record(42,): [{message1}, {message2}],
-            ir.gateway.record(43,): [{message3}],
-            ir.gateway.record(44,): [],
-            ir.gateway.record(): [{messages not handled by any available provider}],
-        }
         """
         result = {}
         remaining_messages = messages[:]
+
+        # Sort providers by sequence (priority)
+        sorted_providers = self.sorted(key=lambda p: p.sequence)
+        _logger.info(
+            "Sorted providers by sequence: %s",
+            [(p.name, p.sequence) for p in sorted_providers],
+        )
+
         while remaining_messages:
             message = remaining_messages.pop()
-            for this in self:
-                if this._can_send(message):
-                    result.setdefault(this, []).append(message)
+            assigned = False
+            for provider in sorted_providers:
+                can_send = provider._can_send(message)
+                _logger.info(
+                    "Provider %s can send to %s: %s",
+                    provider.name,
+                    message.get("number"),
+                    can_send,
+                )
+                if can_send:
+                    result.setdefault(provider, []).append(message)
+                    assigned = True
+                    _logger.info(
+                        "Assigned message %s to provider %s",
+                        message.get("id"),
+                        provider.name,
+                    )
                     break
-            else:
+            if not assigned:
                 result.setdefault(self.browse([]), []).append(message)
+                _logger.info("No provider found for message %s", message.get("id"))
         return result
```

Comment:
In general, could you please try to minimize the diff by avoiding unnecessary changes.

For instance here I don't understand the change in the docstring.
Also the model `ir.sms.gateway` is ordered by sequence, thus I don't understand the sort you are adding.
Finally, how is different the use of your variable `assigned` from the previous `for else` ?